Custom Resource Definition (CRD) and dynamic reload

dnascimento opened the pull request:
**Describe your changes**
This PR adds the following:
- Kubernetes Custom Resource Definition (_powerfulseal.io/v1 scenarios_) to define scenarios. This allows multiple users to define their own scenario. Powerfulseal gets all Scenarios and adds to the scenarios defined in the base configuration file (which is now optional)
- Configuration/Policy file is reloaded after every run (powerfulseal/policy/policy_runner.py)
- Configuration file is now optional, powerfulseal can run only with CRDs.
- Documentation for CRD
- _schemaTools/updateSchema.js_ is a tool to convert the JSON schema to OpenAPIV3 schema (used by CRD). Sorry, I couldn't find one in python.
- Kubernetes: separate sandbox from deployment
- Kubernetes: update sandbox example and add a CRD example

**Testing performed**
WIP. I will commit it them this week

**Additional context**
This aims to make powerfulseal easier to consume by large and multiple teams. The CRD enforces the schema and powerfulseal collates all scenarios before each execution.

Sorry for the large PR

seeker89 commented:
Hey @dnascimento thanks for this PR! I've only had a quick glance now, but I love the idea. 👍 
Let me know when this is ready for a proper review and testing.

dnascimento commented:
@seeker89 I changed it to instantiate PolicyRunner so that it keeps state (configFile and k8sclient).
I liked your static class approach, I had a version with a class PolicyLoader so that PolicyRunner could remain static. I found it more complex and messy than instantiating it. 

I modified the tests to include the CRD k8s invocation. Nevertheless, I structured the code so that it is retro-compatible with existing config files. This version reads the config file and appends the CRD scenarios so users can have a transition period. 

Let me know your thoughts @seeker89  @dgzlopes (I'm also keen to test k6 to complement this test suite)

dgzlopes commented:
Thanks for the PR! Looks great :) 

I'll test it/review it with calm tomorrow.

dgzlopes commented on kubernetes/powerfulseal.yml:
```diff
@@ -6,27 +6,21 @@ metadata:
+      runStrategy:
+        minSecondsBetweenRuns: 30
+        maxSecondsBetweenRuns: 300
+    scenarios: []
```
With `scenarios: []` container keeps restarting (`CrashLoopBackOff`). 

But, If I define a scenario here, all works as expected.

seeker89 replied:
Could you paste the error you're seeing?

dgzlopes replied:
Maybe I'm misunderstanding something (I don't usually run the Seal inside k8's)

So, when I deploy Powerfulseal with this policy.yml:

```yaml
config:
    runStrategy:
      minSecondsBetweenRuns: 5
      maxSecondsBetweenRuns: 10
scenarios: []
```

Powerfulseal pod starts a never-ending loop of `CrashLoopBackOff`:

```
➜  ~ kubectl get pods --namespace=powerfulseal       
NAME                            READY   STATUS              RESTARTS   AGE
powerfulseal-79b8f869b5-nhbdh   0/1     ContainerCreating   0          11s
➜  ~ kubectl --namespace powerfulseal logs powerfulseal-79b8f869b5-nhbdh             
2020-07-08 20:08:59 INFO __main__ verbosity: None; log level: INFO; handler level: INFO
2020-07-08 20:08:59 INFO __main__ Creating kubernetes client with in-cluster config
2020-07-08 20:08:59 INFO k8s_client Initializing with config: None
2020-07-08 20:08:59 INFO __main__ No cloud driver - some functionality disabled
2020-07-08 20:08:59 INFO __main__ Using stdout metrics collector
2020-07-08 20:08:59 INFO __main__ Starting the UI server (0.0.0.0:8000)
2020-07-08 20:08:59 INFO __main__ STARTING AUTONOMOUS MODE
➜  ~ kubectl get pods --namespace=powerfulseal                       
NAME                            READY   STATUS      RESTARTS   AGE
powerfulseal-79b8f869b5-nhbdh   0/1     Completed   2          51s
➜  ~ kubectl --namespace powerfulseal logs powerfulseal-79b8f869b5-nhbdh
2020-07-08 20:09:16 INFO __main__ verbosity: None; log level: INFO; handler level: INFO
2020-07-08 20:09:16 INFO __main__ Creating kubernetes client with in-cluster config
2020-07-08 20:09:16 INFO k8s_client Initializing with config: None
2020-07-08 20:09:16 INFO __main__ No cloud driver - some functionality disabled
2020-07-08 20:09:16 INFO __main__ Using stdout metrics collector
2020-07-08 20:09:16 INFO __main__ Starting the UI server (0.0.0.0:8000)
2020-07-08 20:09:16 INFO __main__ STARTING AUTONOMOUS MODE
```

Later when I install the CDR and deploy the sandbox/scenario everything works fine. That's expected?

seeker89 replied:
This is because of this: https://github.com/bloomberg/powerfulseal/blob/32c9bd796f977b9dde7d312a76164e750c29253d/powerfulseal/policy/policy_runner.py#L82-L83

That condition doesn't make any sense now that we want to be able to run the Seal with no scenarios and then add them later. I think that it can be safely removed.

seeker89 replied:
@dnascimento would you mind removing that conditional and changing the version to `3.0.0` - I'd like to cut the release after adding this. That would be awesome. Thanks so much!

dgzlopes replied:
Makes sense :ok_hand:

seeker89 commented:
I have to say I'm pretty happy with this addition. I think it creates a whole lot of value for little effort, and it's backwards compatible. Thanks @dnascimento !

The `schemaTools/updateSchema.js` is grinding my gears a little bit, but I can't find a good replacement in the python world. I'll try to see what @dgzlopes meant with https://github.com/bloomberg/powerfulseal/pull/281#discussion_r451141042, but otherwise this will be good to go. 👍

dnascimento commented:
Nice catch, that conditional was causing me the same issue! Just updated the version to 3.0.0rc10

dgzlopes commented:
LGTM :rocket:
